source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,344 @@ namespace EnvoyDefault {
 
 using ::envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig;
 using ::Envoy::Http::HeaderString;
+using ::Envoy::Http::Protocol;
+
+HeaderValidator::HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                 StreamInfo::StreamInfo& stream_info)
+    : config_(config), protocol_(protocol), stream_info_(stream_info),
+      header_values_(::Envoy::Http::Headers::get()) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateMethodHeader(const HeaderString& value) {
+  // HTTP Method Registry, from iana.org:
+  // source: https://www.iana.org/assignments/http-methods/http-methods.xhtml
+  //
+  // From the RFC:
+  //
+  // tchar = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" / "."
+  //       /  "^" / "_" / "`" / "|" / "~" / DIGIT / ALPHA
+  // token = 1*tchar
+  // method = token
+  //
+  static absl::node_hash_set<absl::string_view> kHttpMethodRegistry = {
+      "ACL",
+      "BASELINE-CONTROL",
+      "BIND",
+      "CHECKIN",
+      "CHECKOUT",
+      "CONNECT",
+      "COPY",
+      "DELETE",
+      "GET",
+      "HEAD",
+      "LABEL",
+      "LINK",
+      "LOCK",
+      "MERGE",
+      "MKACTIVITY",
+      "MKCALENDAR",
+      "MKCOL",
+      "MKREDIRECTREF",
+      "MKWORKSPACE",
+      "MOVE",
+      "OPTIONS",
+      "ORDERPATCH",
+      "PATCH",
+      "POST",
+      "PRI",
+      "PROPFIND",
+      "PROPPATCH",
+      "PUT",
+      "REBIND",
+      "REPORT",
+      "SEARCH",
+      "TRACE",
+      "UNBIND",
+      "UNCHECKOUT",
+      "UNLINK",
+      "UNLOCK",
+      "UPDATE",
+      "UPDATEREDIRECTREF",
+      "VERSION-CONTROL",
+      "*",
+  };
+
+  const auto& method = value.getStringView();
+  bool is_valid = true;
 
-HeaderValidatorFactory::HeaderValidatorFactory(const HeaderValidatorConfig& config)
-    : config_(config) {}
+  if (config_.restrict_http_methods()) {
+    is_valid = kHttpMethodRegistry.contains(method);
+  } else {
+    is_valid = !method.empty();
+    for (std::size_t i = 0; i < method.size() && is_valid; ++i) {
+      is_valid = testChar(kMethodHeaderCharTable, method.at(i));
+    }
+  }
 
-::Envoy::Http::HeaderValidatorPtr
-HeaderValidatorFactory::create(::Envoy::Http::Protocol protocol,
-                               StreamInfo::StreamInfo& stream_info) {
-  return std::make_unique<HeaderValidator>(config_, protocol, stream_info);
+  if (!is_valid) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidMethod};
+  }
+
+  return HeaderEntryValidationResult::success();
 }
 
-HeaderValidator::HeaderValidator(const HeaderValidatorConfig&, ::Envoy::Http::Protocol,
-                                 StreamInfo::StreamInfo& stream_info)
-    : stream_info_(stream_info) {}
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateSchemeHeader(const HeaderString& value) {
```

Comment:
does current Envoy allow non-http/https schemes?  I respect following the spec but I don't think we want to change Envoy's default behavior here.

Reply:
We've had some back and forth on this. Looking at it again, I think we should just limit scheme to http and https, since this is what we are doing today.